[GSoD'20] Update how the Kubernetes website serves API references (#23294)
kubernetes/website: content/en/docs/concepts/storage/volumes.md

## {{% heading "whatsnext" %}}

Follow an example of [deploying WordPress and MySQL with Persistent Volumes](/docs/tutorials/stateful-application/mysql-wordpress-persistent-volume/).

> Volume is a definition in the Kubernetes REST API.
>   The {{< api-reference page="config-and-storage-resources/volume" >}} object definition describes the object in detail.

Reviewer: ```suggestion
The API reference for `volume` {{< api-reference page="config-and-storage-resources/volume" >}} provides
details about this definition and how is is used in the Kubernetes API.
```
?

Reviewer: _OK to accept as drafted for now, and revise after this is merged_.

Follow an example of [deploying WordPress and MySQL with Persistent Volumes](/docs/tutorials/stateful-application/mysql-wordpress-persistent-volume/).

> The API reference for {{< api-reference page="config-and-storage-resources/volume" >}} provides

Reviewer 2: sigh ...

can we use this syntax for all resource type definitions? How would deal with the following ones?

- `io.k8s.kube-aggregator.pkg.apis.apiregistration/v1`  `ServiceReference`
- `io.k8s.api.admissionregistration/v1beta1` `ServiceReference`
- `io.k8s.apiextensions-apiserver.pkg.apis.apiextensions/v1beta1`  `ServiceReference`

Author: This shortcode only accepts pages references, and pages generally represent top-level resources (with the exception of Container, Volume, ...).

These `ServiceReference` definitions are all inlined, so it is not possible to reference them with this shortcode.

I agree that It would be nice to be able to reference inlined definitions on specific pages, for example the definition of `io.k8s.apiextensions-apiserver.pkg.apis.apiextensions/v1beta1` `ServiceReference` on the page `extend-resources/custom-resource-definition-v1`

Reviewer 3: +1. The pages should be a close interpretation of the spec with a nice presentation.

Reviewer: Because you need a Pod to use a volume, what I'd like (personal opinion) long term is more like:
`{{< api-reference page="workloads-resources/pod-v1" fragment="volume" >}}`

&hellip;does that make sense?

(and to be clear, even if I'd like that long term it doesn't mean I want to hold up merging this change)

Author: > Because you need a Pod to use a volume, what I'd like (personal opinion) long term is more like:
> `{{< api-reference page="workloads-resources/pod-v1" fragment="volume" >}}`
> 
> …does that make sense?
> 
> (and to be clear, even if I'd like that long term it doesn't mean I want to hold up merging this change)

Yes, adding some IDs on the fields, we could reference any of these fields, even if they are nested, like:

`{{< api-reference page="workloads-resources/pod-v1" fragment="dnsConfig.options" >}}`

This could point to, and ideally highlight the inlined description of the `PodDNSConfigOption` object.

![dnsConfig](https://user-images.githubusercontent.com/9973512/99991206-0e77e100-2db5-11eb-9209-58ca0b33234f.png)

Reviewer 2: another question related to the `page` parameter ... why do we have `workloads-resources/pod-v1`?

- does this mean every time I'm referencing the `Pod` API type, I have to know it is a "workload" resource and the path is "workloads-resources"?
- what is the trailing `v1`? is it a shorthand for core v1 group? If that is the design, can you give some examples on other resource types? I'm asking this because I want to know if the `workloads-resources` can be dropped.

This is an important design decision because once we land these pages, we will iterate over it. What is already in will not be easily removed. Even relocating pages may become harmful ... a lot of redirects have to be introduced to keep things working.

Author: The `page` parameter is the path of the target page (removing the common prefix `/docs/reference/kubernetes-api/`).

The `v1` is the version, without the group. For example, the `page` for the `batch/v1` `Job` resource is `workloads-resources/job-v1`.

We consider that there are no **recent** top-level resources with the same name in 2 different groups. **Recent** means that no new version in the same or a different group has deprecated it.

I would be ok to add the group in the URL if you think this assertion is not true or if there is a risk it becomes not true in the future.

Reviewer: @feloy if instead of
https://k8s.io/docs/reference/kubernetes-api/config-and-storage-resources/storage-class-v1/ we had
https://k8s.io/docs/reference/kubernetes-api/storage.k8s.io/storage-class-v1/
(or possibly, even, https://k8s.io/docs/reference/kubernetes-api/storage.k8s.io/storage-class/v1/ with https://k8s.io/docs/reference/kubernetes-api/storage.k8s.io/storage-class/ being a Hugo redirect)

&hellip;do you think that could work?

Even with that change we could _still_ have https://deploy-preview-23294--kubernetes-io-master-staging.netlify.app/docs/reference/kubernetes-api/config-and-storage-resources/ as an index to those resources.

I agree with @tengqm - URL layout is something we're going to commit to for a long time.

Reviewer 3: Hi. Yes, I agree. I am not as interested in the presentation at the moment, however, the **information** that is
presented on the page and how to locate this information is important.

Recently, the page parameter was discussed in regard to adding a shortcode to output the metadata for a page.
I'd like to review that discussion again. There were a number of suggestions.

The issue was locating the correct page and providing a flexible solution. The glossary uses a similar technique to locate the glossary page in a shortcode. I don't think the reference pages should explicitly tie into the partial URL of the page. The root location of the reference pages may not change, but the individual sub-dirs/pages may change.

It was suggested to use "data/headless" pages to contain the metadata and help locate the pages.

Reviewer: Makes sense, with one query:

> I don't think the reference pages should explicitly tie into the partial URL of the page.

what did you mean there by “partial URL” @kbhawkey?

Reviewer 3: The shortcode needs a partial file path to uniquely locate a page?
The author needs to know where the page is located in the file system instead of identifying some parts of the resource
information, `storage.k8s.io/storage-class/v1/`?

Reviewer: I reckon we could (later!) improve the shortcode not to need the partial file path. Does that seem feasible?

If the shortcode can specify the resource type as `storage.k8s.io/v1/storageclass` (or something else that _doesn't_ have the friendly path in it), that's less fragile and easier to document too.

Reviewer 3: 👍

Author: > I reckon we could (later!) improve the shortcode not to need the partial file path. Does that seem feasible?
> 
> If the shortcode can specify the resource type as `storage.k8s.io/v1/storageclass` (or something else that _doesn't_ have the friendly path in it), that's less fragile and easier to document too.

For the moment, the kubernetes-resources-reference generator gets data from the swagger and applies go-templates
 to this data to create Markdown files. Then, these markdown files are copied into the k/website content directory and Hugo handles these files as Markdown files.

In another way, for example for the glossary, "data" files are written into the k/website content directory (at ./content/en/docs/reference/glossary), and Hugo applies templates (at ./layouts/docs/glossary.html) on these data files to create final HTML files.

As I understand, in the first case (with Markdown files), we have a **Branch** Page Bundle of all the API reference Markdown files, and so it is not possible to "search" into this bundle the page for a specific Kubernetes resource.

In another way, with data files as for the glossary, it is possible to have a **Leaf** Page Bundle containing all files under a given directory and search into specific metadata to get the matching page.

So, a solution would be to delegate the final generation to a Hugo layout, instead of making it from the kubernetes-resources-reference. This time, the generation would create HTML directly.

(More on **Leaf/Branch** pages bundles at https://gohugo.io/content-management/page-bundles/)
